core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -1156,6 +1158,28 @@ private[spark] object Utils extends Logging {
   }
 
   /**
+   * Execute a block of code that evaluates to Unit, stop SparkContext is there is any uncaught 
+   * exception
+   *  
+   * NOTE: This method is to be called by the driver-side components to avoid stopping the 
+   * user-started JVM process completely; in contrast, tryOrExit is to be called in the 
+   * spark-started JVM process .
+   */
+  def tryOrStopSparkContext(sc: SparkContext)(block: => Unit) {
+    try {
+      block
+    } catch {
+      case e: ControlThrowable => throw e
+      case t: Throwable =>
+        if (sc != null) {
+          logError(s"uncaught error in thread ${Thread.currentThread().getName}, stopping " +
+            "SparkContext", t)
+          sc.stop()
+        }
```

Comment:
How about throwing `t` again here? So that the user can use `UncaughtExceptionHandler` to monitor the uncaught exception. If not, the user won't be aware that `sc` is shutdown. until calling `runJob` next time.